smart_open/ftp.py

```diff
@@ -0,0 +1,107 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2019 Radim Rehurek <me@radimrehurek.com>
+#
+# This code is distributed under the terms and conditions
+# from the MIT License (MIT).
+#
+
+"""Implements I/O streams over FTP.
+"""
+
+import getpass
+import logging
+import urllib.parse
+import smart_open.utils
+from ftplib import FTP
+
+logger = logging.getLogger(__name__)
+
+SCHEME = "ftp"
+
+"""Supported URL schemes."""
+
+DEFAULT_PORT = 21
+
+URI_EXAMPLES = (
+    "ftp://username@host/path/file",
+    "ftp://username:password@host/path/file",
+    "ftp://username:password@host:port/path/file",
+)
+
+
+def _unquote(text):
+    return text and urllib.parse.unquote(text)
+
+
+def parse_uri(uri_as_string):
+    split_uri = urllib.parse.urlsplit(uri_as_string)
+    assert split_uri.scheme in SCHEME
+    return dict(
+        scheme=split_uri.scheme,
+        uri_path=_unquote(split_uri.path),
+        user=_unquote(split_uri.username),
+        host=split_uri.hostname,
+        port=int(split_uri.port or DEFAULT_PORT),
+        password=_unquote(split_uri.password),
+    )
+
+
+def open_uri(uri, mode, transport_params):
+    smart_open.utils.check_kwargs(open, transport_params)
+    parsed_uri = parse_uri(uri)
+    uri_path = parsed_uri.pop("uri_path")
+    parsed_uri.pop("scheme")
+    return open(uri_path, mode, transport_params=transport_params, **parsed_uri)
+
+
+def convert_transport_params_to_args(transport_params):
+    supported_keywords = [
+        "timeout",
+        "source_address",
+        "encoding",
+    ]
+    unsupported_keywords = [k for k in transport_params if k not in supported_keywords]
+    kwargs = {k: v for (k, v) in transport_params.items() if k in supported_keywords}
+
+    if unsupported_keywords:
+        logger.warning(
+            "ignoring unsupported ftp keyword arguments: %r", unsupported_keywords
+        )
+
+    return kwargs
+
+
+def _connect(hostname, username, port, password, transport_params):
+    kwargs = convert_transport_params_to_args(transport_params)
+    ftp = FTP(**kwargs)
+    try:
+        ftp.connect(hostname, port)
```

Comment:
If an exception was thrown during `connect` we would raise it as an exception during logging in, which seems quite confusing for the user. What do we get out of throwing our own exception here? What are the exceptions for a failed `connect` and a failed `login`?

Reply:
For a failed connect the following error is thrown:
`ConnectionRefusedError: [WinError 10061] No connection could be made because the target machine actively refused it`

For a failed login, the following is thrown:
`ftplib.error_perm: 530 Login incorrect.`

I felt that these errors are somewhat confusing and don't really tell the user the real issue, which is that either their host or port is wrongly entered or a ftp server doesn't exist at that host or port (if connect fails) or the username or password is wrong (if login fails).

Reply:
Those errors are somewhat confusing, but are more informative and specific than anything we could provide. I wouldn't recommend suppressing the exceptions thrown by ftplib

Reply:
Maybe a best of both worlds situation would be
```python3
try:
    ftp.connect(hostname, port)
except Exception e:
   logger.error("Unable to connect to FTP server: try checking the host and port!")
   raise e
try:
    ftp.login(username, password)
except Exception e:
   logger.error("Unable to login to FTP server: try checking the username and password!")
   raise e
```

Reply:
+1 on **not** rephrasing the traceback errors. Because this is what users google. We want them to hit the actual support threads of the core libraries, where people have the same issue – and not smart_open, which is just a wrapper.